### PR TITLE
Additional Config Store Cleanup

### PR DESCRIPTION
**Pull request by gabrieljackson**
This change addresses a few issues where config stores were not
properly closed when an error was encountered on server startup.
This could result in leaked database connections when dealing with
a database config store.

Fixes https://mattermost.atlassian.net/browse/MM-30345

```release-note
Cleanup config store on server initialization errors
```

### CODE REVIEW

**Review by crspeller — APPROVED** [review]
Great!

**Comment by gabrieljackson** [comment]
After a discussion with @crspeller, I confirmed that the config store cannot be (easily) closed once the server is initialized and waiting for a shutdown signal. This would be nice to have as a follow-up in the future, so I figured I would note it here in case there are thoughts/concerns about the idea.

File: config/store.go [hunk 1]
```diff
@@ -54,7 +54,13 @@ func NewStore(dsn string, watch bool, customDefaults *model.Config) (*Store, err
 		return nil, err
 	}
 
-	return NewStoreFromBacking(backingStore, customDefaults)
+	store, err := NewStoreFromBacking(backingStore, customDefaults)
+	if err != nil {
+		backingStore.Close()
+		return store, errors.Wrap(err, "failed to create store")
+	}
```
Comment (agnivade) [inline config/store.go:61]: To make this more correct, we should do `return nil, errors.Wrap(...)`.

File: config/database.go [hunk 2]
```diff
@@ -55,6 +55,7 @@ func NewDatabaseStore(dsn string) (ds *DatabaseStore, err error) {
 		db:             db,
 	}
 	if err = initializeConfigurationsTable(ds.db); err != nil {
+		ds.Close()
```
Comment (agnivade) [inline config/database.go:58]: Since `err` is a named return parameter, we can improve this by moving this to L50 and do a 

```
defer func() {
 if err != nil {
  db.Close()
 }
}()
```

And chance L59 

```
err = errors.Wrap(...)
return nil, err
```

**Review by agnivade — CHANGES_REQUESTED** [review]
Nice fix. I am wondering how hard will it be to write a test for this?

**Review by nronas — COMMENTED** [review]
Nothing more than what @agnivade has already suggested. Some test coverage for this would be awesome. 

Also for my context, will love to know more about the details around the difficulties closing the config when the server has been initialized?

File: cmd/mattermost/commands/server.go [hunk 3]
```diff
@@ -97,6 +97,7 @@ func runServer(configStore *config.Store, usedPlatform bool, interruptChan chan
 	}
 	server, err := app.NewServer(options...)
 	if err != nil {
+		configStore.Close()
```
[the remark this answers is not in the document]
Comment (gabrieljackson) [reply cmd/mattermost/commands/server.go:100]: I was originally avoiding doing double-closes to avoid confusion, but I am happy to change it if you think that is a better approach.

[the remark this answers is not in the document]
Comment (gabrieljackson) [reply cmd/mattermost/commands/server.go:100]: > We also need to call defer a.Srv().Shutdown() after InitDBCommandContext in cmd/mattermost/init.go.

Thank you for this, I figured I was missing properly closing the connections for non-server commands, but hadn't stumbled across where to do this properly.

**Comment by jwilander** [comment]
@gabrieljackson let's also cherry-pick this into `cloud` branch for the cloud release next week

**Comment by gabrieljackson** [comment]
The latest commit contains an important iteration on the fixes for proper server teardown. I wasn't able to include defer'ed shutdown in a common init location, but I double checked all invocations of `InitDBCommandContextCobra` and found two instances where teardown was not called. One of these, `version`, is heavily used in cloud scenarios which probably contributed to the leaked connections were are seeing.

Regarding tests, I took another pass at it and am still struggling to come up with a good way to test that proper teardown is achieved without quite a bit of work refactoring other things. If there is an example you can point me at I would be happy to add tests, but can't find a good sample of how we may properly check this sort of thing.

**Comment by crspeller** [comment]
> This change addresses a few issues where config stores were not
properly closed when an error was encountered on server startup.
This could result in leaked database connections when dealing with
a database config store.

> The latest commit contains an important iteration on the fixes for proper server teardown. I wasn't able to include defer'ed shutdown in a common init location, but I double checked all invocations of InitDBCommandContextCobra and found two instances where teardown was not called. One of these, version, is heavily used in cloud scenarios which probably contributed to the leaked connections were are seeing.

Just a thought. I wonder if this has anything to do with some customers reporting running out of DB connections. FYI @sadohert

File: cmd/mattermost/commands/server.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (agnivade) [reply cmd/mattermost/commands/server.go]: Yep double-closes are typically harmless. But the code looks more readable.

**Comment by agnivade** [comment]
> One of these, version, is heavily used in cloud scenarios which probably contributed to the leaked connections were are seeing.

Nice!

> Regarding tests, I took another pass at it and am still struggling to come up with a good way to test that proper teardown is achieved without quite a bit of work refactoring other things.

Ok, yeah this does seem like a bit of an edge case to test correctly. I believe the challenge would be to cause a failure and return the function early. Checking for the connections is easy enough with a method.

**Review by agnivade — APPROVED** [review]
Nice work on this Gabe

**Comment by agnivade** [comment]
> Based on my new knowledge of this, it basically boils down to the config interface being called throughout the server code so we can't close the original config connections without telling the server to use the main DB connections instead.

So if I am understanding this correctly, during every config update, we want to create a new DB connection, do the update, and then close the connection?

> Edit: perhaps we can ratchet the idle connections down to 1 or 0 for this and/or tweak the idle timeout for the connections?

Ahh you know what, we are not setting `ConnMaxLifetimeMilliseconds` for the db handle that we generate for config/database.go :) It's done in a separate way than our usual sqlstore. Doing that should take care of this concern.

**Comment by nronas** [comment]
> Based on my new knowledge of this, it basically boils down to the config interface being called throughout the server code so we can't close the original config connections without telling the server to use the main DB connections instead.

@crspeller thanks for the context on this. I am a bit confused about this though :D #new-guy, specifically around what config connections are vs main DB connections. My understanding so far is that upon server initialization we create a `sqlStore`, that will create a connection pool configured based on the  `SqlSettings` provided. Based on what I can see in the code when we update the config, and more specifically the SqlSettings, I cannot pinpoint if/where we do the update on the existing connection pool.

**Review by nronas — APPROVED** [review]
LGTM, my comments are for educational purposes on my end :D

**Comment by agnivade** [comment]
> My understanding so far is that upon server initialization we create a sqlStore, that will create a connection pool configured based on the SqlSettings provided.

That's where the confusion is. The config DB backend creates a new connection handle of it's own. Check config/database.go. That's why I commented

> Ahh you know what, we are not setting ConnMaxLifetimeMilliseconds for the db handle that we generate for config/database.go :) It's done in a separate way than our usual sqlstore. Doing that should take care of this concern.

**Comment by nronas** [comment]
> That's where the confusion is. The config DB backend creates a new connection handle of it's own. Check config/database.go. That's why I commented

Thanks @agnivade this is super helpful but also interesting 😄

**Comment by gabrieljackson** [comment]
>  It's done in a separate way than our usual sqlstore. Doing that should take care of this concern.

Exactly, I plan to do a follow-up PR with some tuning for the config DB connections once I see how this change affects our environments.

> What's the primary usage of having that config store use the database since configuration should be ephemeral IMO.

@nronas configuration can be changed in the system console and should be synced across all Mattermost servers when clustered. For this reason and others, database-backed config is an option.

**Comment by nronas** [comment]
> @nronas configuration can be changed in the system console and should be synced across all Mattermost servers when clustered. For this reason and others, database-backed config is an option.

Interesting, so how the remaining nodes(not the one that handled the update) get notified that the config has changed and they have to reload it? Or do they always read from the db when it comes to the config.

**Comment by gabrieljackson** [comment]
@nronas this is handled via server clustering I believe.

**Comment by gabrieljackson** [comment]
Due to the nature of this change, QA testing doesn't seem particularly relevant so we will defer it for now.
